### PR TITLE
Pinecil round two NMSIS

### LINKED ISSUE #730: Updated build/flash instructions for Pinecil?
_This is a Question_

* **I have**
  - [X ] Searched previous issues 
  - [X ] This is in _this_ firmware, not vendor "offical" firmware
  - [ ] This is not a bug in the vendor bootloader (aka DFU)
  - [X ] I have checked this is not already covered in the docs in `/Documentation`

* **What is the current behavior?**

I've tried to build ts100 for Pinecil (both master and the commit that Pinecil shipped with, which is `a6edc9b`), but after flashing it to the Pinecil through dfu it doesn't boot.

Command I've used to flash firmware:
`dfu-util -d 28e9:0189 -a 0 -D ./ci/artifacts/Pinecil_EN.bin -s 0x08000000:mass-erase:force`

And I've built the firmware using build.sh script supplied (basically docker path, where it fetched ubuntu 20.04 container, installed a compiler and built all the firmwares there are).

Interesting enough that I've compared checksum of the stock firmware (made a dump of it and extracted firmware with `dd`, I assume it got shipped with _EN version, but I've also ran the same checks for first 10 KB of firmware which I hope will be similar even if it got shipped not with _EN version) and it's different from what is manually built.

* **What is the expected behavior?**

I expect to be able to build a flashable firmware out of the repository.

***Steps to reproduce the bug:***

https://wiki.pine64.org/wiki/Pinecil#How_to_update_a_firmware - I've tried to document the process for re-flash that I've used there

- ./build.sh
- Wait for it to build
- Press "-" button and connect USB-C plug into Pinecil
- Wait for `dfu-util -l` to detect Pinecil
- Flash firmware with `dfu-util -d 28e9:0189 -a 0 -D ./ci/artifacts/Pinecil_EN.bin -s 0x08000000:mass-erase:force`

* **What are you running:**

On the idle screen, you can hold the settings button and it will show you the firmware & PCB version.
If using a USB-C supply please note model number

  - Firmware Version: 2.12.a6edc9b
  - PCB Version: 1.0
  - Power Supply (Voltage and Current Rating) : N/A

### PR DESCRIPTION
**Pull request by Ralim**
Moving the Pinecil code from the vendor toolchain setup the Nuclei NMSIS setup with their compilers.
Big of hunting to get the Makefile happy and working under Linux.
Fair bit of reshuffling to do that, so open to any comments or ideas :)

This should fix #730 and from my testing the CI builds and my local builds work for me :)
There was a fair bit of going around in circles to get here, apologies for the delay, but didn't want to publish a branch of non-functional code.

TODO:

- [x] Verify no regressions in Miniware Irons
- [x] Verify no regressions in Pinecil

### CODE REVIEW

**Comment by Firebie** [comment]
On windows I get next error:
`riscv-nuclei-elf-g++.exe: error: Objects/Pinecil/./Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/GCC/startu: No such file or directory`

**Comment by Ralim** [comment]
Is this using WSL or how are you running it under windows ?

**Comment by Firebie** [comment]
Non-WSL, pure windows.
Yes,  somehow filename is cut or somehow name is cut from 'startup_gd32vf103.o'

**Comment by Firebie** [comment]
As for master branch - it compiles successfully.

**Comment by Ralim** [comment]
I wonder if this is a limit on windows with command line length?

I have a vague memory of windows having this as a limiting factor.

The number of arguments have grown longer so this could make sense.

**Comment by Firebie** [comment]
`Linking Pinecil_EN.elf
-Wl,--gc-sections -Wl,--wrap=malloc -Wl,--wrap=free -oHexfile/Pinecil_EN.elf -Wl,-Map=Hexfile/Pinecil_EN.map -lm -Wl,--undefined=vTaskSwitchContext -Wl,--undefined=pxCurrentTCB -Wl,--defsym=__FLASH_SIZE__=128k -Wl,--defsym=__BOOTLDR_SIZE__=0x0 -lstdc++ -nostartfiles -Xlinker --gc-sections --specs=nosys.specs -u _isatty -u _write -u _sbrk -u _read -u _close -u _fstat -u _lseek --specs=nano.specs
riscv-nuclei-elf-g++.exe: error: Objects/Pinecil/./Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/GCC/startu: No such file or directory
Makefile:285: recipe for target 'Hexfile/Pinecil_EN.elf' failed
make: *** [Hexfile/Pinecil_EN.elf] Error 1
    [Error]
*********************************************
 -- Stop on error --`

**Comment by Firebie** [comment]
For me it looks like issue is from 'riscv-nuclei-elf-g++.exe' - it can't handle the path

**Comment by Firebie** [comment]
It looks like compiler issue, it can't handle path length > 80 chars:
Objects/Pinecil/./Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/GCC/startu

**Comment by Firebie** [comment]
Renaming 'gd32vf103' folder to 'g' allows to pass this point

**Comment by Firebie** [comment]
Did some research - windows is limited to 8191 for command line length. Current linker command line size is around 8700 chars.
Can we re-organize some folders so the linker will work on windows too?

**Comment by Ralim** [comment]
Windows is by itself _not_ a supported OS, but under docker is and WSL is best-effort.

That said, open to ideas of re-organisation to help with this.

Is the issue just the path to the `ld` file? I'm slowly moving all of the device specific files into the BSP subfolder to keep the main "core" logic cleanly separated from the model specific mess.
Could just place a symlink to get around the issues in the linker if required?

**Comment by Firebie** [comment]
The issue on windows that total size of command for linker is higher than 8191 - [test.txt.zip](https://github.com/Ralim/ts100/files/5686257/test.txt.zip)

File: workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h [hunk 1]
```diff
@@ -57,7 +55,7 @@ extern uint32_t SystemCoreClock;
 
 /* Software timer related definitions. */
 #define configUSE_TIMERS 0
-#define configTIMER_TASK_PRIORITY (configMAX_PRIORITIES - 1) //Tmr_svc ç‹¬å� æœ€é«˜ä¼˜å…ˆçº§
+#define configTIMER_TASK_PRIORITY3
```
Comment (Cl1608Ho) [inline workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h:58]: Did i spot a missing whitespace in front of the 3?

File: workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h [hunk 2]
```diff
@@ -1,21 +1,18 @@
 #ifndef FREERTOS_CONFIG_H
 #define FREERTOS_CONFIG_H
 /* Ensure stdint is only used by the compiler, and not the assembler. */
-#if defined(__ICCARM__) || defined(__CC_ARM) || defined(__GNUC__)
```
Comment (Cl1608Ho) [inline workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h:4]: This comment appears to be related to that #if you removed... might remove the comment too, or state what exactly is happening?

**Comment by Cl1608Ho** [comment]
Thanks for the work ralim.... i took a quick glance over it and while i cannot comment on any on the actually functional stuff (1. not experienced, 2. its a bit of a mess of code changes + formatting changes so its hard to see what is "just" formatting changes and where code actually changed), i did notice those two (maybe?) style issues, hope it helps!

**Comment by Firebie** [comment]
If I rename 'Pine64' to 'P' and 'gd32vf103' to 'g', compilation is passed successfully.

File: workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h (hunk 1 above)
Comment (Civil) [reply workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h:58]: I wonder what's the reason behind changing `(configMAX_PRIORITIES - 1)` to `3` (as previously it was more clear that timer task should have priority 1 below highest.

**Comment by Civil** [comment]
@Cl1608Ho when viewing the diff in github web view you can click on settings button and tick "Hide whitespace" changes and that would hide all whitespace-only changes and that would reduce noise.

There was a flag for `diff` or `git diff` that I always forgets :)

File: workspace/TS100/Core/BSP/Pine64/I2C_Wrapper.cpp [hunk 3]
```diff
@@ -10,523 +10,532 @@
...
+			} else {
+				timeout = 0;
+				state = I2C_START;
+				in_rx_cycle = 0;
+			}
+			break;
+		case I2C_CLEAR_ADDRESS_FLAG:
+			/* address flag set means i2c slave sends ACK */
+			while ((!i2c_flag_get(I2C0, I2C_FLAG_ADDSEND))
+					&& (timeout < I2C_TIME_OUT )) {
```
Comment (Civil) [inline workspace/TS100/Core/BSP/Pine64/I2C_Wrapper.cpp:108]: Trailing whitespace after `I2C_TIME_OUT` before `))`? (Actually happens later as well, in same context)

File: workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_spi.c [hunk 4]
```diff
@@ -2,7 +2,7 @@
     \file  gd32vf103_spi.c
     \brief SPI driver
 
-    \version 2019-06-05, V1.0.0, firmware for GD32VF103
+    \version 2019-6-5, V1.0.0, firmware for GD32VF103
```
Comment (Civil) [inline workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_spi.c:5]: This looks like some automated changes across all files.

File: workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c [hunk 5]
```diff
@@ -980,7 +979,7 @@ void timer_channel_output_fast_config(uint32_t timer_periph, uint16_t channel, u
     \param[in]  channel:
                 only one parameter can be selected which is shown as below:
       \arg        TIMER_CH_0: TIMER channel 0(TIMERx(x=0..4))
-      \arg        TIMER_CH_1: TIMER channel 1(TIMERx(x=0..4))
+      \arg        TIMER_CH_1: TIMER channel 1(TIMERx(x=0..41))
```
Comment (Civil) [inline workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c:982]: I do not really understand those changes to the comments. I mean in general in this file.

I was trying to check those things with the https://files.pine64.org/doc/datasheet/pinecil/GD32VF103_Datasheet_Rev%201.1.pdf and it looks like that original comments were correct everywhere in the file (GD32VF103 have 4 generic timers and not 41).

File: workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c [hunk 6]
```diff
@@ -1964,3 +1963,5 @@ void timer_flag_clear(uint32_t timer_periph, uint32_t flag)
 {
     TIMER_INTF(timer_periph) = (~(uint32_t)flag);
 }
+
```
Comment (Civil) [inline workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c:1966]: Trailing new lines.

**Comment by Civil** [comment]
btw, I've tried to build a firmware from this branch and it works even if it was built under Linux.

File: workspace/TS100/Core/BSP/Pine64/I2C_Wrapper.cpp (hunk 3 above)
Comment (Ralim) [reply workspace/TS100/Core/BSP/Pine64/I2C_Wrapper.cpp:108]: To be honest I'm just using the default formatter in the Nuclei IDE (eclipse) so unsure why the extra space is there.. hmm.

File: workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_spi.c (hunk 4 above)
Comment (Ralim) [reply workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_spi.c:5]: I have changed from using the drivers published on the gigadevice website to the ones from NMSIS

Moved to this one : https://github.com/Nuclei-Software/nuclei-sdk

File: workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c (hunk 5 above)
Comment (Ralim) [reply workspace/TS100/Core/BSP/Pine64/Vendor/SoC/gd32vf103/Common/Source/Drivers/gd32vf103_timer.c:982]: I have changed from using the drivers published on the gigadevice website to the ones from NMSIS

Moved to this one : https://github.com/Nuclei-Software/nuclei-sdk

So I didnt do any of these changes, but rather just moved to their version of the files. (all the gd32vf103_* files are direct from that repo)

**Comment by Ralim** [comment]
@Civil 
What is the original issue you have not resolved?

**Comment by Civil** [comment]
My original problem (why I wanted to try updated firmware or at least to build my own) is that Pinecil cannot communicate with any PD charger (so I can get a QC 3.0, but not a PD, and I've tried about 4 or 5 different charges and a pack of type-c cables with different PD ratings). So my plan was to build a firmware for it and hope that it will fix my issue. Or the second option - to solder break-out board, connect jtag and hope that I will be able to get some idea of what's failing.

**Comment by Ralim** [comment]
Hmm that is very interesting.

Feels more likely to be hardware than software though.

When connected to a PD power supply do you just get 5v? 

Assuming your PD supplies are rated 45W+ you should definitely get 20V.

Could build a firmware to debug the PD state maybe.

**Comment by Civil** [comment]
Yup, I only get 5V out of PD charger. And I've tried several of them: 35W Panasonic one (supports 12V, but not 20), 87W from Apple, 65W laptop charger from Lenovo, PD Charger from Pixel phone, and some random QC+PD charger that I bought just for the pinecil (100W rated, that supports all voltages that are in the standard).

**Comment by Ralim** [comment]
Yeah, should definitely get something more than 5v out.
FUSB302 could be funky or not talking. 

Been thinking of adding the of state to the debug menu, to test you could hack this in and see where it's sitting?

**Comment by Civil** [comment]
Yeah, I'll try to hack something in. It's though been a while since I've done any embedded development :)

I actually suspect that CC1 or CC2 pins on the FUSB302 is not soldered correctly on mine. Another symptom is that if charger supports PowerDelivery (and only then) Pinecil won't turn on in one position of the connector, but it will work fine if charger only supports QC.

**Comment by Ralim** [comment]
Ah, yes that sounds like an extremely plausible reason for the issue. Definitely sounds like a lifted CC connection :'(

**Comment by Pine-A64** [comment]
@Civil, I am PINE64 info team and just read the log. Please submit a support ticket at https://desk.zoho.com/portal/pine64/home​ and mentioned to support team that info team refer you to file this support ticket. The support team will help up issue a RMA on Pinecil replacement.

File: workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h (hunk 1 above)
Comment (Ralim) [reply workspace/TS100/Core/BSP/Pine64/FreeRTOSConfig.h]: Not entirely sure, but at this point of trying to get something out before 🎄 , deferring to the nuclei settings.

**Comment by Civil** [comment]
@Pine-A64 that was a first thing I did (Ticker number is 8039), but while I was waiting for a reply I've decided to try to diagnose it myself. Also it seems support was fairly busy with the requests :)

**Comment by Firebie** [comment]
About compiling under windows - I found solution using compiler option to read command line options from file:
diff --git i/workspace/TS100/Makefile w/workspace/TS100/Makefile
index 9f0e98e3..abead31f 100644
--- i/workspace/TS100/Makefile
+++ w/workspace/TS100/Makefile
@@ -286,7 +286,8 @@ all: $(OUT_HEXFILE).hex $(OUT_HEXFILE).bin
 $(OUT_HEXFILE).elf : $(OUT_OBJS_S) $(OUT_OBJS) $(OUT_OBJS_CPP) Makefile $(LDSCRIPT)
 	@test -d $(@D) || mkdir -p $(@D)
 	@echo Linking $(OUTPUT_EXE).elf
-	@$(CPP) $(CXXFLAGS) $(OUT_OBJS_S) $(OUT_OBJS) $(OUT_OBJS_CPP) $(LIBS) $(LINKER_FLAGS)
+  @echo $(OUT_OBJS) > Objects/OUT_OBJS.txt
+	@$(CPP) $(CXXFLAGS) $(OUT_OBJS_S) @Objects/OUT_OBJS.txt $(OUT_OBJS_CPP) $(LIBS) $(LINKER_FLAGS)
 
 $(OUT_OBJS): $(OUTPUT_DIR)/%.o : %.c Makefile
 	@test -d $(@D) || mkdir -p $(@D)

**Pull request merged**